src/lib/comms/sol-socket-impl-zephyr.c

```diff
@@ -0,0 +1,423 @@
+ */
+
+#include <errno.h>
+
+#include <net/ip_buf.h>
+#include <sections.h>
+
+#include "sol-buffer.h"
+#include "sol-log.h"
+#include "sol-mainloop.h"
+#include "sol-mainloop-zephyr.h"
+#include "sol-socket-impl.h"
+#include "sol-vector.h"
+#include "sol-network-util.h"
+
+struct sol_socket_net_context;
+
+struct sol_socket_zephyr {
+    struct sol_socket base;
+
+    struct {
+        bool (*cb)(void *data, struct sol_socket *s);
+        const void *data;
+    } read, write;
+
+    struct sol_timeout *write_timeout;
+
+    struct sol_socket_net_context *unicast_context;
+    struct sol_ptr_vector mcast_contexts;
+    struct sol_ptr_vector bufs;
+
+    struct nano_sem lock;
+    bool read_available;
+};
+
+#define RECV_STACKSIZE 256
+
+struct sol_socket_net_context {
+    struct sol_socket_zephyr *socket;
+    struct net_context *context;
+    nano_thread_id_t fiber;
+    uint8_t stack[RECV_STACKSIZE] __stack;
+};
+
+static void
+socket_read_available(void *data)
+{
+    struct sol_socket_zephyr *s = data;
+    uint16_t count;
+
+    nano_task_sem_take(&s->lock, TICKS_UNLIMITED);
+    count = sol_ptr_vector_get_len(&s->bufs);
+    s->read_available = false;
+    nano_task_sem_give(&s->lock);
+
+    if (s->read.cb)
+        while (count--)
+            s->read.cb((void *)s->read.data, &s->base);
+}
+
+static void
+socket_signal_mainloop(struct sol_socket_zephyr *s, struct net_buf *buf)
+{
+    struct mainloop_event me = {
+        .cb = socket_read_available,
+        .data = s
+    };
+    int ret;
+
+    nano_fiber_sem_take(&s->lock, TICKS_UNLIMITED);
+
+    ret = sol_ptr_vector_append(&s->bufs, buf);
+    SOL_INT_CHECK_GOTO(ret, < 0, err);
+
+    if (!s->read_available) {
+        sol_mainloop_event_post(&me);
+        s->read_available = true;
+    }
+
+err:
+    nano_fiber_sem_give(&s->lock);
+}
+
+static void
+socket_recv_fiber(int arg1, int arg2)
+{
+    struct sol_socket_net_context *ctx = (struct sol_socket_net_context *)arg1;
+
+    while (1) {
+        struct net_buf *buf;
+
+        buf = net_receive(ctx->context, TICKS_UNLIMITED);
+        if (buf)
+            socket_signal_mainloop(ctx->socket, buf);
+    }
+}
+
+static struct sol_socket *
+sol_socket_zephyr_new(int domain, enum sol_socket_type type, int protocol)
+{
+    struct sol_socket_zephyr *socket;
+
+    SOL_INT_CHECK_GOTO(domain, != SOL_NETWORK_FAMILY_INET6,
+        unsupported_family);
+
+    socket = calloc(1, sizeof(*socket));
+    SOL_NULL_CHECK_GOTO(socket, socket_error);
+
+    sol_ptr_vector_init(&socket->mcast_contexts);
+    sol_ptr_vector_init(&socket->bufs);
+    nano_sem_init(&socket->lock);
+    nano_sem_give(&socket->lock);
+
+    return &socket->base;
+
+socket_error:
+    errno = ENOMEM;
+    return NULL;
+
+unsupported_family:
+    errno = EAFNOSUPPORT;
+    return NULL;
+}
+
+static void
+socket_net_context_del(struct sol_socket_net_context *ctx)
+{
+    if (ctx->fiber)
+        fiber_fiber_delayed_start_cancel(ctx->fiber);
```

Comment:
Please add a comment explaining why using this function here is ... less than ideal.